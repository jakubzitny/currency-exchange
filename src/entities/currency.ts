// export type Currency = 'USD' | 'GBP' | 'EUR'

export enum Currency {
  USD = 'USD',
  GBP = 'GBP',
  EUR = 'EUR',
}
