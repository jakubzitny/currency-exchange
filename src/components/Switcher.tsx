import React from 'react'
import classNames from 'classnames'
import { List } from 'immutable'
import { v4 } from 'uuid'

import { Currency } from '../entities/currency'
import Wallet from '../entities/wallet'

import styles from '../styles/Switcher.module.css'

type Props = {
  wallets: List<Wallet>
  activeWallet: Currency
  type: 'source' | 'target'
  onSwitchRequest: (currency: Currency) => void
}

export default function Switcher(props: Props) {
  const handleSwitchClick = (
    currency: Currency,
    e: React.MouseEvent<HTMLDivElement, MouseEvent>
  ) => {
    props.onSwitchRequest(currency)
  }

  return (
    <div
      className={classNames(styles.container, {
        [styles.containerTarget]: props.type === 'target',
      })}
    >
      {props.wallets.toArray().map((wallet) => (
        <div
          key={v4()}
          className={classNames(styles.switchDot, {
            [styles.switchDotActive]:
              wallet.get('currency') === props.activeWallet,
          })}
          onClick={handleSwitchClick.bind(null, wallet.get('currency'))}
        />
      ))}
    </div>
  )
}
