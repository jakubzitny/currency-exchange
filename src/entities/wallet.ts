import { Record } from 'immutable'
import { Currency } from './currency'

interface Defaults {
  currency: Currency
  amount: number
}

const defaults: Defaults = {
  currency: Currency.EUR,
  amount: 0,
}

export default class Wallet extends Record<Defaults>(defaults) {}
