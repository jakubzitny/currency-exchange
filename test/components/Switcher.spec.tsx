import chai, { expect } from 'chai'
import chaiSinon from 'sinon-chai'
import chaiJestSnapshot from 'chai-jest-snapshot'
import { shallow } from 'enzyme'
import React from 'react'
import { spy } from 'sinon'
import { List } from 'immutable'

import { Currency } from '../../src/entities/currency'
import Wallet from '../../src/entities/wallet'
import Switcher from '../../src/components/Switcher'

chai.use(chaiJestSnapshot)
chai.use(chaiSinon)

describe('Switcher', () => {
  const createSwitcher = (
    props: Object = {},
    type: 'source' | 'target' = 'source'
  ) => {
    const wallet = new Wallet({ currency: Currency.EUR, amount: 0 })
    const wallet2 = new Wallet({ currency: Currency.USD, amount: 0 })
    const defaultProps = {
      wallets: List.of(wallet, wallet2),
      activeWallet: Currency.EUR,
      onSwitchRequest: () => {},
    }
    const switcher = <Switcher type={type} {...defaultProps} {...props} />
    return shallow(switcher)
  }

  it('should render source Switcher', () => {
    const wrapper = createSwitcher()
    expect(wrapper).to.matchSnapshot()
  })

  it('should render target Switcher', () => {
    const wrapper = createSwitcher({}, 'target')
    expect(wrapper).to.matchSnapshot()
  })

  it('should handle wallet switch click', () => {
    const onSwitchRequest = spy()
    const wrapper = createSwitcher({ onSwitchRequest })

    wrapper.find('.switchDot').at(1).simulate('click')
    expect(onSwitchRequest).to.have.been.called
  })
})
