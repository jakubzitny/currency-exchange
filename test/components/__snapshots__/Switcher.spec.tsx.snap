// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switcher should render source Switcher 1`] = `ShallowWrapper {}`;

exports[`Switcher should render target Switcher 1`] = `ShallowWrapper {}`;
